### ISSUE TITLE
`clippy::nonminimal_bool` False positive on `!0u32`

### ISSUE THREAD

**Issue by ithinuel** [rust-lang/rust-clippy#12408]
### Summary

Clippy suggests to transform `var == !0u32` into `var != 0u32` which are not equivalent.

See: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=97011897f0ad26ee23256f928eae9df2

### Lint Name

clippy::nonminimal_bool

### Reproducer

I tried this code:

```rust
use std::hint::black_box;

fn dev_id() -> u32 {
    0xFFFF_FFFF
}

fn main() {
    let id = black_box(dev_id());
    if id == 0 || id == !0 {
        println!("yes");
    }
}
```

I saw this happen:

```
warning: this boolean expression can be simplified
 --> src/main.rs:9:19
  |
9 |     if id == 0 || id == !0 {
  |                   ^^^^^^^^ help: try: `id != 0`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#nonminimal_bool
  = note: `#[warn(clippy::nonminimal_bool)]` on by default
```

I expected to see this happen:
No warnings.

### Version

```text
rustc 1.78.0-nightly (2dceda4f3 2024-03-01)
binary: rustc
commit-hash: 2dceda4f32b97f60b122f2b32491e0267ef5cc0c
commit-date: 2024-03-01
host: x86_64-unknown-linux-gnu
release: 1.78.0-nightly
LLVM version: 18.1.0
```

**Comment by y21** [comment]
This was fixed in #12372 and the reproducer does not emit a warning on master anymore, so I think it's safe to close this :)

**Comment by ithinuel** [comment]
Oh, I didn’t see that, thanks!
